python/tskit/stats.py

```diff
@@ -163,3 +167,457 @@ def r2_matrix(self):
+            self.sample_sets = [sample_sets]
+        else:
+            assert all([isinstance(i, list) for i in sample_sets])
+            assert all(
+                [
+                    all([isinstance(j, (int, np.int32, np.int64)) for j in i])
+                    for i in sample_sets
+                ]
+            )
+            self.sample_sets = sample_sets
+        assert all([i in ts.samples() for j in self.sample_sets for i in j])
+
+        if weight_func is None:
+
+            def _count_events(node, tree, sample_sets):
+                return np.array([1])
+
+            self.weight_func = _count_events
+        elif weight_func == "pair_coalescence_events":
+            self.weight_func = self._count_pair_coalescence_events
+        else:
+            assert callable(weight_func)
+            self.weight_func = weight_func
+        _weight_func_eval = self.weight_func(0, ts.first(), self.sample_sets)
+        assert isinstance(_weight_func_eval, np.ndarray)
+        assert _weight_func_eval.ndim == 1
+        self.num_weights = len(_weight_func_eval)
+
+        if window_breaks is None:
+            window_breaks = np.array([0.0, ts.sequence_length])
+        else:
+            assert isinstance(window_breaks, np.ndarray)
+            assert window_breaks.ndim == 1
+            assert np.min(window_breaks) >= 0.0
+            assert np.max(window_breaks) <= ts.sequence_length
+            window_breaks = np.sort(np.unique(window_breaks))
+        self.windows = [
+            trees.Interval(left, right)
+            for left, right in zip(window_breaks[:-1], window_breaks[1:])
+        ]
+        self.num_windows = len(self.windows)
+
+        if blocks_per_window is None:
+            self.num_blocks = 1
+        else:
+            assert isinstance(blocks_per_window, int)
+            assert blocks_per_window > 0
+            self.num_blocks = blocks_per_window
+
+        self.tables = [table for table in self._generate_ecdf_tables(ts, window_breaks)]
+
+    # TODO add __str__ and __repr_html__ methods
+    #
+    # def __str__(self):
+    #    return self.useful_text_summary()
+    #
+    # def __repr_html__(self):
+    #    return self.useful_html_summary()
+
+    def copy(self):
+        return copy.deepcopy(self)
+
+    def ecdf(self, times):
+        """
+        Returns the empirical distribution function evaluated at the time
+        points in ``times``.
+
+        The output array has shape ``(self.num_weights, len(times),
+        self.num_windows)``.
+        """
+
+        assert isinstance(times, np.ndarray)
+        assert times.ndim == 1
+
+        values = np.zeros((self.num_weights, len(times), self.num_windows))
+        for k, table in enumerate(self.tables):
+            indices = np.searchsorted(table.time, times, side="right")
+            values[:, :, k] = np.array(
+                [
+                    [0.0] * self.num_weights
```

Comment:
I think this could be an intermediate value assigned to a varianble? There's a lot of logic happening in here. Also probably more efficient to do things with numpy indexing rather than go through in a list comprehension.